factor/totp/classes/factor.php

```diff
@@ -184,19 +184,31 @@ public function login_form_definition($mform) {
     public function login_form_validation($data) {
         $factors = $this->get_active_user_factors();
         $result = array('verificationcode' => get_string('error:wrongverification', 'factor_totp'));
+        $windowconfig = get_config('factor_totp', 'window');
 
         foreach ($factors as $factor) {
             $totp = TOTP::create($factor->secret);
-
+            // Convert seconds to windows.
+            $window = (int) floor($windowconfig / $totp->getPeriod());
             // First check if this code matches the last verified timestamp.
             $lastverified = $this->get_lastverified($factor->id);
-            if ($totp->verify($data['verificationcode'], $lastverified, 1)) {
+            if ($totp->verify($data['verificationcode'], $lastverified, $window)) {
                 return array('verificationcode' => get_string('error:codealreadyused', 'factor_totp'));
             }
 
-            if ($totp->verify($data['verificationcode'], time(), 1)) {
-                $result = array();
+            $pasttimestamp = time() - (5 * $totp->getPeriod());
+            $futuretimestamp = time() + (5 * $totp->getPeriod());
+            $time = userdate(time()) . ' ' . usertimezone();
```

Comment:
timezone should be in the format eg strftime('%F %T %Z'); and throw that into a new lang string like

langconfig.php:46:$string['strftimedaydate'] = '%A, %d %B %Y';

Reply:
note we can tune the string so it is just focused on the time and timezone, we don't care about the date really